### ISSUE TITLE
'expected 2-element int8 array' error  while getting data from continuous aggregate (Using query with subquery)

### ISSUE THREAD

**Issue by anjithapaulose** [timescale/timescaledb#2693]
I am having table with mac_address and uptime. Mac address will be given with colon or without colon. I am having the following data

```
   mac_address   |      timestamp      | uptime
-------------------+---------------------+----------
 AA:BB:45:TT:CC:34 | 2020-11-30 00:00:04 |    0
 112233445566      | 2020-11-30 00:00:06 |    0
 NN:CC:33:33:22:22 | 2020-11-30 00:00:05 |    0
 AA:BB:45:TT:CC:34 | 2020-11-30 00:30:04 |    1
 112233445566      | 2020-11-30 00:30:06 |    1
 NN:CC:33:33:22:22 | 2020-11-30 00:30:05 |    1 
```

I have created an hourly continous agggregate view to get average uptime in each hour using below query

```
CREATE VIEW aggregated_data
WITH (timescaledb.continuous) AS
SELECT
time_bucket(INTERVAL '1 hour', timestamp) AS time_bucket,
mac_address,
AVG(uptime) AS avg_uptime
FROM data_table
GROUP BY time_bucket, mac_address
```

When I am trying to get data from aggregated_data table using below query, I am getting error '**expected 2-element int8 array**'

```
SELECT mac_address, time_bucket,
avg_uptime, count(*) over() as count
FROM aggregated_data
WHERE time_bucket >= '2020-10-28 18:00:00' and time_bucket < '2020-11-30 08:31:52'
AND mac_address IN (
          SELECT mac_address
          FROM aggregated_data
          WHERE time_bucket >= '2020-11-28 18:00:00' and time_bucket < '2020-11-30 08:31:52'
          GROUP BY mac_address
          HAVING AVG(avg_uptime) > 0 AND AVG(avg_uptime) <=10
  )
ORDER BY time_bucket LIMIT 10 OFFSET 0
```

**Comment by k-rus** [comment]
@anjithapaulose Thank you for reporting. What version of PostgreSQL do you use? What is the version of TimesaleDB, `\dx`?

I tried to reproduce the issue using the above information but without success. Can you provide more details about your example, including the table definition and minimal amount of data to insert?
What is the result of `\d+ data_table`?

Can you provide the result of `\d+ aggregated_data`?

**Comment by anjithapaulose** [comment]
We are getting the data from view after creation of the continuous aggregate view. But when new data is added to data_table, we are not able to get data view(aggregated_data)

**Postgres version** : 12.5
**Timescale db version**: 1.7.4

**aggregated_data**
`SELECT _materialized_hypertable_2.time_bucket,
 _materialized_hypertable_2.mac_address,
 _timescaledb_internal.finalize_agg('avg(integer)'::text, NULL::name, NULL::name, '{{pg_catalog,int4}}'::name[], _materialized_hypertable_2.agg_2_2, NULL::numeric) AS avg_uptime,
 FROM _timescaledb_internal._materialized_hypertable_2
 WHERE _materialized_hypertable_2.time_bucket < COALESCE(_timescaledb_internal.to_timestamp_without_timezone(_timescaledb_internal.cagg_watermark('2'::oid)), '-infinity'::timestamp without time zone)
 GROUP BY _materialized_hypertable_6.time_bucket, _materialized_hypertable_6.mac_address
 UNION ALL
 SELECT time_bucket('01:00:00'::interval, data_table."timestamp") AS time_bucket,
 data_table.mac_address
 avg(data_table.uptime) AS avg_uptime
 FROM data_table
 WHERE data_table."timestamp" >= COALESCE(_timescaledb_internal.to_timestamp_without_timezone(_timescaledb_internal.cagg_watermark('2'::oid)), '-infinity'::timestamp without time zone)
 GROUP BY (time_bucket('01:00:00'::interval, data_table."timestamp")), data_table.mac_address;`

**data_table**
`mac_address            | character varying(20)       |  
 timestamp               | timestamp without time zone |
 uptime                    | integer                     |`

**Comment by k-rus** [comment]
@anjithapaulose Thank you providing more details. Unfortunately it is not enough to reproduce the issue.
Based on the above information I created the following repo:
```sql
create table data_table (
    mac_address character varying(20),
    timestamp timestamp without time zone not null,
    uptime integer);
select create_hypertable('data_table','timestamp');
insert into data_table values 
('AA:BB:45:TT:CC:34', '2020-11-30 00:00:04', 0),
('112233445566', '2020-11-30 00:00:06', 0),
('NN:CC:33:33:22:22', '2020-11-30 00:00:05', 0),
('AA:BB:45:TT:CC:34', '2020-11-30 00:30:04', 1),
('112233445566', '2020-11-30 00:30:06', 1),
('NN:CC:33:33:22:22', '2020-11-30 00:30:05', 1);

CREATE VIEW aggregated_data
WITH (timescaledb.continuous) AS
SELECT
time_bucket(INTERVAL '1 hour', timestamp) AS time_bucket,
mac_address,
AVG(uptime) AS avg_uptime
FROM data_table
GROUP BY time_bucket, mac_address;

SELECT mac_address, time_bucket,
avg_uptime, count(*) over() as count
FROM aggregated_data
WHERE time_bucket >= '2020-10-28 18:00:00' and time_bucket < '2020-11-30 08:31:52'
AND mac_address IN (
          SELECT mac_address
          FROM aggregated_data
          WHERE time_bucket >= '2020-11-28 18:00:00' and time_bucket < '2020-11-30 08:31:52'
          GROUP BY mac_address
          HAVING AVG(avg_uptime) > 0 AND AVG(avg_uptime) <=10
  )
ORDER BY time_bucket LIMIT 10 OFFSET 0;
```
Executing the query returns a result and don't give any error:
```
    mac_address    |     time_bucket     |       avg_uptime       | count
-------------------+---------------------+------------------------+-------
 112233445566      | 2020-11-30 00:00:00 | 0.50000000000000000000 |     3
 AA:BB:45:TT:CC:34 | 2020-11-30 00:00:00 | 0.50000000000000000000 |     3
 NN:CC:33:33:22:22 | 2020-11-30 00:00:00 | 0.50000000000000000000 |     3
(3 rows)
```
So it looks to be a some corner case.
Can you write a similar repro script, which results in the error?

**Comment by pehlert** [comment]
Hi all,

We are experiencing the same issue and can reproduce it, but I had absolutely no luck at all pinpointing what it might be.

This is the query we are running:

```
SELECT
	avg(speed) AS speed
	FROM continuous_events_speed
	INNER JOIN inventory as machine ON (machine.guid = continuous_events_speed.guid)
	INNER JOIN license_machine as lm on lm.machine_guid = machine.guid
	INNER JOIN license on license.id = lm.license_id
	WHERE TRUE
	AND	machine.location_id = '1'
	AND bucket >= '2021-02-01 00:00:00.000 +00:00'
	AND bucket <= '2021-02-28 23:59:59.999 +00:00'        
	AND license."type" = 'BASIC'        
	AND date(license.license_end) > '2021-03-11'        
	AND date(license.license_start) <= '2021-03-11'
	--AND license.license_end > '2021-03-11'        
	--AND license.license_start <= '2021-03-11'
	;```
	
	
You see the date() function there above the commented lines. When we use date, it fails with the error message:

```
Query 1 ERROR: ERROR:  expected 2-element int8 array
```

Without the date() it works just fine.

This also only fails when we attempt to load data from the `continuous_events_speed` table, which is a continuous aggregate:

```
 SELECT _materialized_hypertable_12.guid,
    _materialized_hypertable_12.bucket,
    _timescaledb_internal.finalize_agg('avg(integer)'::text, NULL::name, NULL::name, '{{pg_catalog,int4}}'::name[], _materialized_hypertable_12.agg_3_3, NULL::numeric)::integer AS speed
   FROM _timescaledb_internal._materialized_hypertable_12
  WHERE _materialized_hypertable_12.bucket < COALESCE(_timescaledb_internal.to_timestamp_without_timezone(_timescaledb_internal.cagg_watermark(12)), '-infinity'::timestamp without time zone)
  GROUP BY _materialized_hypertable_12.guid, _materialized_hypertable_12.bucket
UNION ALL
 SELECT events.guid,
    time_bucket('1 day'::interval, events."timestamp") AS bucket,
    avg(events.speed)::integer AS speed
   FROM events
  WHERE events.stop::text = 'STOP_NO'::text AND events."timestamp" >= COALESCE(_timescaledb_internal.to_timestamp_without_timezone(_timescaledb_internal.cagg_watermark(12)), '-infinity'::timestamp without time zone)
  GROUP BY events.guid, (time_bucket('1 day'::interval, events."timestamp"));```
  
  
It also only happens for data for a specific range (bucket) of data. This fails:

```SELECT
	avg(speed) AS speed
	FROM continuous_events_speed
	INNER JOIN inventory as machine ON (machine.guid = continuous_events_speed.guid)
	INNER JOIN license_machine as lm on lm.machine_guid = machine.guid
	INNER JOIN license on license.id = lm.license_id
	WHERE TRUE
	AND	machine.location_id = '1'
	AND bucket >= '2021-02-01 00:00:00.000 +00:00'
	AND bucket <= '2021-02-06 23:59:59.999 +00:00'        
	AND license."type" = 'BASIC'        
	AND date(license.license_end) > '2021-03-11'        
	AND date(license.license_start) <= '2021-03-11'
	--AND license.license_end > '2021-03-11'        
	--AND license.license_start <= '2021-03-11'
	;```

While this works:

```SELECT
	avg(speed) AS speed
	FROM continuous_events_speed
	INNER JOIN inventory as machine ON (machine.guid = continuous_events_speed.guid)
	INNER JOIN license_machine as lm on lm.machine_guid = machine.guid
	INNER JOIN license on license.id = lm.license_id
	WHERE TRUE
	AND	machine.location_id = '1'
	AND bucket >= '2021-02-01 00:00:00.000 +00:00'
	AND bucket <= '2021-02-05 23:59:59.999 +00:00'        
	AND license."type" = 'BASIC'        
	AND date(license.license_end) > '2021-03-11'        
	AND date(license.license_start) <= '2021-03-11'
	--AND license.license_end > '2021-03-11'        
	--AND license.license_start <= '2021-03-11'
	;```

And this also works:

```
SELECT
	avg(speed) AS speed
	FROM continuous_events_speed
	INNER JOIN inventory as machine ON (machine.guid = continuous_events_speed.guid)
	INNER JOIN license_machine as lm on lm.machine_guid = machine.guid
	INNER JOIN license on license.id = lm.license_id
	WHERE TRUE
	AND	machine.location_id = '1'
	AND bucket >= '2021-02-02 00:00:00.000 +00:00'
	AND bucket <= '2021-02-07 23:59:59.999 +00:00'        
	AND license."type" = 'BASIC'        
	AND date(license.license_end) > '2021-03-11'        
	AND date(license.license_start) <= '2021-03-11'
	--AND license.license_end > '2021-03-11'        
	--AND license.license_start <= '2021-03-11'
	;
	```


I am very happy to provide more information, but will need some guidance to be able to find a pattern here I think.

**Comment by pehlert** [comment]
@anjithapaulose Could you please check whether this issue still occurs when you do `set enable_hashagg to false;`?

**Comment by pehlert** [comment]
cc @svenklemm This seemingly unrelated issue was also solved for me by disabling HashAgg

**Comment by svenklemm** [comment]
@pehlert can you provide a minimal script that reproduces the issue in a fresh database

**Comment by pehlert** [comment]
I wish I could @svenklemm ,  but no matter what I tried, I could not find any pattern to this. I could however attempt with my existing database and the patch you recently pushed for the gapfill issue. Would that help?

**Comment by svenklemm** [comment]
That bugfix shouldnt affect this query since it was for gapfill specifically.

**Comment by svenklemm** [comment]
@pehlert can you post EXPLAIN of the failing query i'm guessing simple explain works and the error happens during execution.

**Comment by arnevanacker** [comment]
Hi,
I have the same problem. Setting the enable_hashagg to false, does indeed 'fix' this.

My hypertable:
> ts - timestamp without time zone
> sensor - text
> f - integer
> i - integer
> u - integer
> p - integer
> s - integer
> pf - integer
> tstz - timestamp with time zone
> Indexes:
>     "electrical_data_sensor_ts_tstz_key" UNIQUE CONSTRAINT, btree (sensor, ts, tstz)
> Foreign-key constraints:
>     "electrical_data_sensor_fkey" FOREIGN KEY (sensor) REFERENCES sensors(sn)
> Triggers:
>     ts_cagg_invalidation_trigger AFTER INSERT OR DELETE OR UPDATE ON electrical_data FOR EACH ROW EXECUTE FUNCTION _timescaledb_internal.continuous_agg_invalidation_trigger('1')
>     ts_insert_blocker BEFORE INSERT ON electrical_data FOR EACH ROW EXECUTE FUNCTION _timescaledb_internal.insert_blocker()

My cont_agg:
> bucket - timestamp without time zone
> sensor - text
> sum_p  - bigint
> avg_p  - numeric
> max_p  - integer
> min_p  - integer
> sum_s  - bigint
> avg_s  - numeric
> max_s  - integer
> min_s  - integer

My view:
> bucket - timestamp without time zone
> energy - numeric
> peak - bigint
> power - numeric
> node - integer
> project - integer
> include_for_total_consumption - boolean
> View definition:
>  SELECT a.bucket,
>     sum(a.sum_p) AS energy,
>     sum(a.max_s) AS peak,
>     sum(a.avg_p) AS power,
>     n.id AS node,
>     n.project,
>     n.include_for_total_consumption
>    FROM aggregates_5m a
>      JOIN nodes n ON (n.sensor_l1 = a.sensor OR n.sensor_l2 = a.sensor OR n.sensor_l3 = a.sensor) AND n.start <= a.bucket AND a.bucket < n.finish
>   GROUP BY a.bucket, n.id, n.project, n.include_for_total_consumption;

Like others before have posted. Some queries work while others don't.
This query gives 'ERROR:  expected 2-element int8 array' as output
> SELECT time_bucket('5M',bucket)::text as x, trunc(sum(power)/1000.0,1) as y FROM nodes_agg_5m
> WHERE bucket >= '2021-05-02T22:00:00.000Z'
> AND bucket < '2021-05-03T22:00:00.000Z'
> AND node = 11 and project = 95 GROUP BY x ORDER BY x;
The same query but for data one day earlier works fine.
Or if you replace 'node=11' with 'include_for_total_consumption' the query will also work.

Anything more I can do to help figure out what happens?

**Comment by NunoFilipeSantos** [comment]
@pehlert can you please come back to us with the EXPLAIN outcome?

**Comment by pehlert** [comment]
Sorry for the delay. Here it comes:

```
Aggregate  (cost=55.69..55.70 rows=1 width=32)
  ->  Nested Loop  (cost=7.58..55.68 rows=1 width=4)
        ->  Nested Loop  (cost=7.44..55.21 rows=1 width=78)
"              Join Filter: ((lm.machine_guid)::text = (""*SELECT* 1"".guid)::text)"
              ->  Hash Join  (cost=1.35..2.60 rows=1 width=37)
                    Hash Cond: (lm.license_id = licenses.id)
                    ->  Seq Scan on license_machine lm  (cost=0.00..1.19 rows=19 width=41)
                    ->  Hash  (cost=1.34..1.34 rows=1 width=4)
                          ->  Seq Scan on licenses  (cost=0.00..1.34 rows=1 width=4)
"                                Filter: ((type = 'BASIC'::enum_license_type) AND (date(license_end) > '2021-03-11'::date) AND (date(license_start) <= '2021-03-11'::date))"
              ->  Append  (cost=6.09..51.93 rows=54 width=41)
"                    ->  Subquery Scan on ""*SELECT* 1""  (cost=6.09..7.31 rows=49 width=41)"
                          ->  HashAggregate  (cost=6.09..6.82 rows=49 width=49)
                                Group Key: _materialized_hypertable_12.guid, _materialized_hypertable_12.bucket
                                ->  Custom Scan (ChunkAppend) on _materialized_hypertable_12  (cost=0.00..5.72 rows=49 width=90)
                                      Chunks excluded during startup: 0
                                      ->  Seq Scan on _hyper_12_3172_chunk  (cost=0.00..5.72 rows=49 width=90)
"                                            Filter: ((bucket >= '2021-02-01 00:00:00'::timestamp without time zone) AND (bucket <= '2021-02-28 23:59:59.999'::timestamp without time zone) AND (bucket < COALESCE(_timescaledb_internal.to_timestamp_without_timezone(_timescaledb_internal.cagg_watermark(12)), '-infinity'::timestamp without time zone)))"
"                    ->  Subquery Scan on ""*SELECT* 2""  (cost=44.16..44.35 rows=5 width=41)"
                          ->  GroupAggregate  (cost=44.16..44.30 rows=5 width=49)
"                                Group Key: events.guid, (time_bucket('1 day'::interval, events.""timestamp""))"
                                ->  Sort  (cost=44.16..44.17 rows=5 width=49)
"                                      Sort Key: events.guid, (time_bucket('1 day'::interval, events.""timestamp""))"
                                      ->  Custom Scan (ChunkAppend) on events  (cost=0.29..44.10 rows=5 width=49)
                                            Chunks excluded during startup: 5
        ->  Index Scan using inventory_guid_idx on inventory machine  (cost=0.14..0.46 rows=1 width=37)
"              Index Cond: ((guid)::text = (""*SELECT* 1"".guid)::text)"
              Filter: (location_id = 1)
```

**Comment by arnevanacker** [comment]
Just FYI: I can indeed execute explain query, but not explain analyse.
I'm on timescaleDB 12.2 on postgres 12.

**Comment by svenklemm** [comment]
Found a reproducer for this while investigating different bug:
```
CREATE TABLE metrics(filler_1 int, time timestamptz NOT NULL, device_id int, v0 int, v1 int, v2 float, v3 float);
SELECT create_hypertable('metrics','time',create_default_indexes:=false);

ALTER TABLE metrics DROP COLUMN filler_1;
INSERT INTO metrics(time,device_id,v0,v1,v2,v3) SELECT time, device_id, device_id+1,  device_id + 2, device_id + 0.5, NULL FROM generate_series('2000-01-01 0:00:00+0'::timestamptz,'2000-01-05 23:55:00+0','2m') gtime(time), generate_series(1,5,1) gdevice(device_id);


create materialized view cagg1 WITH (timescaledb.continuous,timescaledb.materialized_only=true) AS SELECT time_bucket('1h',time), device_id, min(v0), max(v0), avg(v0) FROM metrics GROUP BY 1,2;
select from metrics as m, lateral(select m from cagg1 where avg is NULL limit 1) as lat;
```

**Comment by svenklemm** [comment]
This was fixed with the cagg changes introduced in 2.7.0

**Comment by pehlert** [comment]
Glad to hear! We'll reenable hash agg after the upgrade and keep observing!
